torchgen/gen_functionalization_type.py

```diff
@@ -353,6 +339,132 @@ def emit_view_functionalization_body(
 """
 
 
+def maybe_create_output(f: NativeFunction, var_name: str) -> str:
+    if len(f.func.returns) == 0:
+        return ""
+    return_type = dispatcher.returns_type(f.func.returns).remove_const_ref().cpp_type()
+    return f"{return_type} {var_name} = "
+
+
+def wrap_propagate_mutations_and_return(
+    f: NativeFunction, functional_op: Optional[NativeFunction], inner_out_name: str
+) -> str:
+    if len(f.func.returns) > 1:
+        # Some assertions: We don't want any functions with a return type of "-> (Tensor(a!), Tensor)",
+        # because:
+        # (1) It's more annoying to handle properly
+        # (2) It's unnecessary - you can't method-chain on the first (mutated) output because it's part of a tuple.
+        # Instead, we expect the (a!) argument to not be returned.
+        mutable_rets = [
+            r
+            for r in f.func.returns
+            if r.annotation is not None and r.annotation.is_write
+        ]
+        immutable_rets = [r for r in f.func.returns if r.annotation is None]
```

Comment:
This smacks of "we need to make the returns model" more structured. Having to manually parse out semantic structure at the end use site suggests the model isn't refined enough.